### PR TITLE
Inbox capacity bounding by type

### PR DESCRIPTION
**Pull request by mzabolotko**

### CODE REVIEW

File: .paket/Paket.Restore.targets [hunk 1]
```diff
@@ -43,7 +43,7 @@
     <!-- Step 1 Check if lockfile is properly restored -->
     <PropertyGroup>
       <PaketRestoreRequired>true</PaketRestoreRequired>
-      <NoWarn>$(NoWarn);NU1603</NoWarn>
+      <NoWarn>$(NoWarn);NU1603;NU1604;NU1605;NU1608</NoWarn>
```
Comment (alexzaitzev) [inline .paket/Paket.Restore.targets:46]: Remove it, if it was added unintended

File: Kontur.sln [hunk 2]
```diff
@@ -21,7 +21,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Kontur.Rabbitmq", "Sources\
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Kontur.Rabbitmq.Tests", "Tests\Kontur.Rabbitmq.Tests\Kontur.Rabbitmq.Tests.csproj", "{68E20730-598D-4ED8-9C6E-F099BE325653}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Kontur.Rabbitmq.IntegrationTests", "Tests\Kontur.Rabbitmq.IntegrationTests\Kontur.Rabbitmq.IntegrationTests.csproj", "{B7886B1F-8577-4CB2-B927-824AE8013307}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Kontur.Rabbitmq.IntegrationTests", "Tests\Kontur.Rabbitmq.IntegrationTests\Kontur.Rabbitmq.IntegrationTests.csproj", "{B7886B1F-8577-4CB2-B927-824AE8013307}"
```
Comment (alexzaitzev) [inline Kontur.sln:24]: Looks like these changes also unintended, better to revert

File: .paket/Paket.Restore.targets (hunk 1 above)
Comment (alexzaitzev) [inline .paket/Paket.Restore.targets:46]: Undo these changes, if you not intended to do it.

File: Sources/Kontur.Rabbitmq/AmqpPublishingBuilder.cs [hunk 3]
```diff
@@ -7,8 +7,8 @@ namespace Kontur.Rabbitmq
     public class AmqpPublishingBuilder : IAmqpPublishingBuilder
     {
         private readonly IAmqpRouter router;
-        private readonly IAmqpPropertyBuilder propertyBuilder;
-        private readonly List<Func<IAmqpMessageBuilder, IAmqpConnectionFactory, IPublisher>> publishers;
+        private readonly AmqpPropertyBuilder propertyBuilder;
```
Comment (alexzaitzev) [inline Sources/Kontur.Rabbitmq/AmqpPublishingBuilder.cs]: Implementation should be under the interfaces

File: Sources/Kontur.Rabbitmq/AmqpPublishingBuilder.cs [hunk 4]
```diff
@@ -7,8 +7,8 @@ namespace Kontur.Rabbitmq
     public class AmqpPublishingBuilder : IAmqpPublishingBuilder
     {
         private readonly IAmqpRouter router;
-        private readonly IAmqpPropertyBuilder propertyBuilder;
-        private readonly List<Func<IAmqpMessageBuilder, IAmqpConnectionFactory, IPublisher>> publishers;
+        private readonly AmqpPropertyBuilder propertyBuilder;
+        private readonly List<Func<IAmqpMessageBuilder, IAmqpConnectionFactory, IPublisherRegistry, IPublishingTag>> publishers;
 
         public IAmqpConnectionFactory ConnectionFactory { get; private set; }
```
Comment (alexzaitzev) [inline Sources/Kontur.Rabbitmq/AmqpPublishingBuilder.cs]: Private set is redundant

File: Sources/Kontur/Bus.cs [hunk 5]
```diff
@@ -4,57 +4,86 @@
...
 namespace Kontur
 {
     public class Bus : IPublisherRegistry, ISubscriptionRegistry
     {
-        private readonly BufferBlock<IMessage> inbox = new BufferBlock<IMessage>();
-        private readonly BroadcastBlock<IMessage> dispatcher = new BroadcastBlock<IMessage>(message => message);
-        private readonly ConcurrentDictionary<string, ISubscriptionTag> subscribers = new ConcurrentDictionary<string, ISubscriptionTag>();
-
-        public Bus()
+        private readonly ConcurrentDictionary<Type, MessageBuffer> inboxes =
```
Comment (alexzaitzev) [inline Sources/Kontur/Bus.cs]: It's better for readability to split declaration and initialization.

File: Sources/Kontur/ISubscriptionRegistry.cs [hunk 6]
```diff
@@ -5,8 +5,8 @@ namespace Kontur
     public interface ISubscriptionRegistry
     {
         bool IsSubscribed(ISubscriptionTag tag);
-        ISubscriptionTag Subscribe<T>(Action<Message<T>> subscriber);
-        ISubscriptionTag Subscribe<T>(ISubscriber subscriber);
+        ISubscriptionTag Subscribe<T>(Action<Message<T>> subscriber, int queueCapacity = 1);
```
Comment (alexzaitzev) [inline Sources/Kontur/ISubscriptionRegistry.cs:8]: In future we should place it into separate class with default value in it

File: Sources/Kontur/MessageDispatcher.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+using System.Threading.Tasks.Dataflow;
+
+using MessageTarget = System.Threading.Tasks.Dataflow.ITargetBlock<Kontur.IMessage>;
+using MessageTargetDictionary = System.Collections.Concurrent.ConcurrentDictionary<string, System.Threading.Tasks.Dataflow.ITargetBlock<Kontur.IMessage>>;
+
+namespace Kontur
+{
+    internal class MessageDispatcher
+    {
+        ConcurrentDictionary<Type, MessageTargetDictionary> routes =
```
Comment (alexzaitzev) [inline Sources/Kontur/MessageDispatcher.cs]: Property can be made `readonly` and split declaration also

File: Sources/Kontur/MessageDispatcher.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        private bool Unsubscribe<T>(string id)
+        {
+            if (routes.TryGetValue(typeof(T), out MessageTargetDictionary value))
+            {
+                if (value.TryRemove(id, out MessageTarget target))
+                {
+                    target.Complete();
+                    return true;
+                }
+                else
```
Comment (alexzaitzev) [inline Sources/Kontur/MessageDispatcher.cs]: Redundant else

File: Sources/Kontur/MessageDispatcher.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                {
+                    return false;
+                }
+            }
+            else
```
Comment (alexzaitzev) [inline Sources/Kontur/MessageDispatcher.cs]: Redundant else

File: Sources/Kontur/MessageDispatcher.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+            }
+        }
+
+        internal int GetCountSubscriberOf(Type type)
+        {
+            if (routes.TryGetValue(type, out MessageTargetDictionary value))
+            {
+                return value.Count;
+            }
+            else
```
Comment (alexzaitzev) [inline Sources/Kontur/MessageDispatcher.cs]: Redundant else

File: Sources/Kontur/MessageDispatcherTag.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+
+namespace Kontur
+{
+    internal class MessageDispatcherTag : IDisposable
```
Comment (alexzaitzev) [inline Sources/Kontur/MessageDispatcherTag.cs:5]: Looks that we have too many tags, maybe we can extract interface `Itag` and make only one class `Tag`, for example

File: Tests/Kontur.Rabbitmq.Tests/AmqpPublishingBuilderFixture.cs [hunk 12]
```diff
@@ -37,7 +37,7 @@ public void CanBuildWithoutPublishers()
 
             IPublishingTag publishingTag = sut.Build(registry);
 
-            A.CallTo(() => registry.RegisterPublisher(A<IPublisher>.Ignored)).MustNotHaveHappened();
+            A.CallTo(() => registry.RegisterPublisher<int>(A<IPublisher>.Ignored)).MustNotHaveHappened();
```
Comment (alexzaitzev) [inline Tests/Kontur.Rabbitmq.Tests/AmqpPublishingBuilderFixture.cs]: Trailing spaces at the end of the line

File: Tests/Kontur.Tests/MessageDispatcherFixture.cs [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            A.CallTo(() => message.RouteKey).Returns(typeof(int));
+
+            var sut = new MessageDispatcher();
+            sut.Subscribe<int>(target);
+            sut.Subscribe<char>(target);
+            sut.Subscribe<char>(target);
+            sut.Subscribe<string>(target);
+            sut.Subscribe<double>(target);
+
+
```
Comment (alexzaitzev) [inline Tests/Kontur.Tests/MessageDispatcherFixture.cs]: Extra new line

File: Tests/Kontur.Tests/MessageDispatcherFixture.cs [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            Task task = sut.Dispatch(message);
+            task.IsCompleted.Should().BeTrue();
+            target.Count.Should().Be(1);
+        }
+
```
Comment (alexzaitzev) [inline Tests/Kontur.Tests/MessageDispatcherFixture.cs]: Extra line

File: Tests/Kontur.Tests/MessageDispatcherFixture.cs [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+            sut.Subscribe<int>(target);
+            sut.Subscribe<int>(target);
+
+            Task task = sut.Dispatch(message);
+            task.IsCompleted.Should().BeTrue();
+            target.Count.Should().Be(2);
+        }
+
+        [Test]
+        public void CanDispatchToSomeSubscribedType()
```
Comment (alexzaitzev) [inline Tests/Kontur.Tests/MessageDispatcherFixture.cs]: Name does not reflect what test does

**Review by alexzaitzev — CHANGES_REQUESTED** [review]
Please, look at the comments and make changes accordingly.

File: .paket/Paket.Restore.targets (hunk 1 above)
Comment (mzabolotko) [reply .paket/Paket.Restore.targets:46]: I don't control this file. It updates automatically.

File: Kontur.sln (hunk 2 above)
Comment (mzabolotko) [reply Kontur.sln:24]: I have changed the type of the visual studio project to .net core. I think it is O.K.

File: .paket/Paket.Restore.targets (hunk 1 above)
Comment (mzabolotko) [reply .paket/Paket.Restore.targets:46]: I don't control this file. It updates automatically.

File: Sources/Kontur.Rabbitmq/AmqpPublishingBuilder.cs (hunk 3 above)
Comment (mzabolotko) [reply Sources/Kontur.Rabbitmq/AmqpPublishingBuilder.cs]: Fixed.

File: Sources/Kontur.Rabbitmq/AmqpPublishingBuilder.cs (hunk 4 above)
Comment (mzabolotko) [reply Sources/Kontur.Rabbitmq/AmqpPublishingBuilder.cs]: Fixed

File: Sources/Kontur/MessageDispatcherTag.cs (hunk 11 above)
Comment (mzabolotko) [reply Sources/Kontur/MessageDispatcherTag.cs:5]: https://github.com/mzabolotko/Kontur/issues/47

File: Tests/Kontur.Tests/MessageDispatcherFixture.cs (hunk 13 above)
Comment (mzabolotko) [reply Tests/Kontur.Tests/MessageDispatcherFixture.cs]: Fixed.

File: Tests/Kontur.Tests/MessageDispatcherFixture.cs (hunk 14 above)
Comment (mzabolotko) [reply Tests/Kontur.Tests/MessageDispatcherFixture.cs]: Fixed.

File: Tests/Kontur.Rabbitmq.Tests/AmqpPublishingBuilderFixture.cs (hunk 12 above)
Comment (mzabolotko) [reply Tests/Kontur.Rabbitmq.Tests/AmqpPublishingBuilderFixture.cs]: Fixed.

File: Tests/Kontur.Tests/MessageDispatcherFixture.cs (hunk 15 above)
Comment (mzabolotko) [reply Tests/Kontur.Tests/MessageDispatcherFixture.cs]: Fixed.

File: Sources/Kontur/Bus.cs (hunk 5 above)
Comment (mzabolotko) [reply Sources/Kontur/Bus.cs]: Fixed.

File: Sources/Kontur/MessageDispatcher.cs (hunk 7 above)
Comment (mzabolotko) [reply Sources/Kontur/MessageDispatcher.cs]: Fixed.

File: Sources/Kontur/MessageDispatcher.cs (hunk 8 above)
Comment (mzabolotko) [reply Sources/Kontur/MessageDispatcher.cs]: Fixed

File: Sources/Kontur/MessageDispatcher.cs (hunk 10 above)
Comment (mzabolotko) [reply Sources/Kontur/MessageDispatcher.cs]: Fixed

File: Sources/Kontur/ISubscriptionRegistry.cs (hunk 6 above)
Comment (mzabolotko) [reply Sources/Kontur/ISubscriptionRegistry.cs:8]: O.K.

**Review by alexzaitzev — APPROVED** [review]
LGFM

**Pull request merged**